File: .github/scripts/update_discord.py
import requests
import re
import os

# API URL
API_URL = "https://api.sangnguyen07.io.vn/api/service/discord"

def get_discord_data():
    response = requests.get(API_URL)
    if response.status_code == 200:
        return response.json()
    else:
        print(f"Error fetching Discord data: {response.status_code}")
        return None

def get_status_badge(status):
    status_colors = {
        "online": "green",
        "idle": "yellow",
        "dnd": "red",
        "offline": "gray"
    }
    color = status_colors.get(status, "gray")
    return f"<img src=\"https://img.shields.io/badge/status-{status}-{color}?style=flat-square\" alt=\"{status.upper()} Status\">"

def get_badges_text(badges):
    badge_names = {
        "HypeSquadOnlineHouse1": "HypeSquad Bravery",
        "HypeSquadOnlineHouse2": "HypeSquad Brilliance",
        "HypeSquadOnlineHouse3": "HypeSquad Balance",
        "ActiveDeveloper": "Active Developer",
        "Nitro": "Nitro",
        "NitroClassic": "Nitro Classic"
    }
    return ", ".join([badge_names.get(badge, badge) for badge in badges])

def format_activity(activity):
    details = activity.get("details", "")
    state = activity.get("state", "")
    
    activity_text = f"<b>Activity:</b> {activity.get('name', '')}<br>"
    if details:
        activity_text += f"<i>{details}</i><br>"
    if state:
        activity_text += f"{state}"
    
    return activity_text

def generate_discord_section(data):
    username = data.get("username", "")
    displayname = data.get("displayname", "")
    avatar = data.get("avatar", "")
    status = data.get("status", "offline")
    badges = data.get("badges", [])
    roles = data.get("roles", [])
    activities = data.get("activities", [])
    
    # Find the highest role (usually the first one with color)
    main_role = None
    for role in roles:
        if role.get("name") != "@everyone" and role.get("color") != "#000000":
            main_role = role
            break
    
    # Get activity if present
    activity_text = ""
    if activities:
        activity_text = format_activity(activities[0])
    
    section = f"""### 🎮 Discord Presence

<div align="center">
  <img src="https://img.shields.io/badge/Discord-{username}-5865F2?style=for-the-badge&logo=discord&logoColor=white" alt="Discord">
  <p><b>Status:</b> {get_status_badge(status)}</p>
  
  <table>
    <tr>
      <td rowspan="4">
        <img src="{avatar}" width="100" alt="Discord Avatar">
      </td>
      <td><b>Username:</b> {username} ({displayname})</td>
    </tr>
    <tr>
      <td><b>Badges:</b> {get_badges_text(badges)}</td>
    </tr>"""
    
    if main_role:
        section += f"""
    <tr>
      <td><b>Role:</b> <span style="color:{main_role.get('color')}">{main_role.get('name')}</span></td>
    </tr>"""
    
    section += f"""
    <tr>
      <td>
        {activity_text}
      </td>
    </tr>
  </table>
</div>

<!-- Thông tin này được cập nhật tự động từ API: https://api.sangnguyen07.io.vn/api/service/discord -->"""
    
    return section

def update_readme():
    discord_data = get_discord_data()
    if not discord_data:
        return
    
    discord_section = generate_discord_section(discord_data)
    
    # Read the current README.md
    with open("README.md", "r", encoding="utf-8") as file:
        content = file.read()
    
    # Check if Discord section already exists
    discord_pattern = r"### 🎮 Discord Presence.*?<!-- Thông tin này được cập nhật tự động từ API.*?-->"
    
    if re.search(discord_pattern, content, re.DOTALL):
        # Replace existing section
        updated_content = re.sub(discord_pattern, discord_section, content, flags=re.DOTALL)
    else:
        # Add section before the profile views or at the end
        if "---\n<p align=" in content:
            updated_content = content.replace("---\n<p align=", f"{discord_section}\n\n---\n<p align=")
        else:
            updated_content = f"{content}\n\n{discord_section}"
    
    # Write updated content back to README.md
    with open("README.md", "w", encoding="utf-8") as file:
        file.write(updated_content)
    
    print("README.md updated with Discord presence")

if __name__ == "__main__":
    update_readme()

File: .github/scripts/update_discord.js
// update_discord.js - Script để cập nhật thông tin Discord trong README.md
const fs = require('fs');
const path = require('path');
// Sử dụng node-fetch v2 để tương thích với môi trường GitHub Actions
const fetch = require('node-fetch');

// Thông tin API
const DISCORD_ID = "920508073423028274";
const API_BASE = "https://api.sangnguyen07.io.vn";
const API_ENDPOINT = "/api/service/discord";

// Badge map cho Discord
const discordBadges = {
  "HypeSquadOnlineHouse1": "https://cdn.discordapp.com/badge-icons/8a88d63823d8a71cd5e390baa45efa02.png", // Bravery
  "HypeSquadOnlineHouse2": "https://cdn.discordapp.com/badge-icons/011940fd013da3f7fb926e4a1cd2e618.png", // Brilliance
  "HypeSquadOnlineHouse3": "https://cdn.discordapp.com/badge-icons/3aa41de486fa12454c3761e8e223442e.png", // Balance
  "ActiveDeveloper": "https://cdn.discordapp.com/badge-icons/6bdc42827a38498929a4920da12695d9.png"
};

// Fetch Discord data từ API sử dụng fetch như đề xuất
async function fetchDiscordData() {
  try {
    console.log("Fetching Discord data...");
    
    const response = await fetch("".concat(API_BASE, API_ENDPOINT), {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        discordId: DISCORD_ID
      })
    });
    
    console.log(`Status Code: ${response.status}`);
    
    if (!response.ok) {
      console.error(`Failed with status code: ${response.status}`);
      return null;
    }
    
    const data = await response.json();
    return data;
  } catch (error) {
    console.error("Error fetching Discord data:", error);
    return null;
  }
}

// Lấy badge cho trạng thái
function getStatusBadge(status) {
  const statusColors = {
    "online": "green",
    "idle": "yellow", 
    "dnd": "red",
    "offline": "gray"
  };
  
  const statusEmojis = {
    "online": "🟢",
    "idle": "🟡",
    "dnd": "🔴",
    "offline": "⚫"
  };
  
  const color = statusColors[status] || "gray";
  const emoji = statusEmojis[status] || "⚫";
  const statusText = status.toUpperCase();
  
  return `<img src="https://img.shields.io/badge/${emoji}_${statusText}-${color}?style=flat-square" alt="Status">`;
}

// Format hoạt động hiện tại
function formatActivity(activities) {
  if (!activities || activities.length === 0) {
    return "Không có hoạt động";
  }
  
  for (const activity of activities) {
    if (activity.type === 0) { // Đang chơi game
      const name = activity.name || "";
      const details = activity.details || "";
      const state = activity.state || "";
      
      let activityText = `<b>${name}</b><br>`;
      if (details) {
        activityText += `<i>${details}</i><br>`;
      }
      if (state) {
        activityText += `${state}`;
      }
      
      return activityText;
    }
  }
  
  return "Đang hoạt động";
}

// Xử lý các ký tự đặc biệt trong role name cho URL
function encodeRoleName(name) {
  return encodeURIComponent(name).replace(/-/g, "--").replace(/%/g, "-").replace(/\./g, "-");
}

// Tạo phần Discord cho README
function generateDiscordSection(data) {
  // Lấy thông tin cơ bản
  const username = data.username || "sang0023";
  const displayName = data.displayname || username;
  const avatarUrl = data.avatar || "https://cdn.discordapp.com/embed/avatars/0.png";
  const userId = data.id || DISCORD_ID;
  
  // Lấy trạng thái online
  const status = data.status || "offline";
  
  // Lấy hoạt động hiện tại
  const activities = data.activities || [];
  const activityText = formatActivity(activities);
  
  // Lấy role chính
  const roles = data.roles || [];
  let mainRole = null;
  
  for (const role of roles) {
    if (role.name !== "@everyone" && role.color !== "#000000") {
      mainRole = role;
      break;
    }
  }
  
  // Bắt đầu tạo HTML
  let discordHtml = `### 🎮 Discord 

<div align="center">
  <img src="https://img.shields.io/badge/Discord-${username}-5865F2?style=for-the-badge&logo=discord&logoColor=white" alt="Discord">
  <img src="https://discord-readme-badge.vercel.app/api?id=${userId}" alt="Discord Presence" />
</div>

<table align="center">
  <tr>
    <td align="center" width="150px" valign="top">
      <img width="100" height="100" src="${avatarUrl}">
      <br>
      ${getStatusBadge(status)}
    </td>
    <td align="left">
      <h3>${displayName} <code>#${username}</code></h3>`;
      
  if (mainRole) {
    const roleColor = mainRole.color || "#e91e63";
    const roleName = mainRole.name || "OWNER";
    const encodedRoleName = encodeRoleName(roleName);
    const colorHex = roleColor.replace('#', '');
    
    discordHtml += `
      <p>
        <img src="https://img.shields.io/badge/${encodedRoleName}-${colorHex}?style=flat-square" alt="Role">
      </p>`;
  } else {
    discordHtml += `
      <p>
        <img src="https://img.shields.io/badge/OWNER-e91e63?style=flat-square" alt="Role">
      </p>`;
  }
  
  // Xử lý badges
  const badges = data.badges || [];
  if (badges.length > 0) {
    discordHtml += `
      <p>`;
    
    for (const badge of badges) {
      if (discordBadges[badge]) {
        discordHtml += `
        <img src="${discordBadges[badge]}" width="20" height="20" alt="${badge}">`;
      }
    }
    
    discordHtml += `
      </p>`;
  }
  
  discordHtml += `
      <h4>📱 ĐANG HOẠT ĐỘNG</h4>
      <blockquote>
        ${activityText}
      </blockquote>
    </td>
  </tr>
</table>

<!-- Cập nhật vào: ${new Date().toLocaleString('vi-VN')} -->
<!-- Thông tin này được cập nhật tự động từ API: ${API_BASE}${API_ENDPOINT} -->`;

  return discordHtml;
}

// Cập nhật README
async function updateReadme() {
  try {
    const discordData = await fetchDiscordData();
    
    if (!discordData) {
      console.log("No Discord data available. Using static content.");
      return;
    }
    
    console.log("Generating Discord section...");
    const discordSection = generateDiscordSection(discordData);
    
    const readmePath = path.join(process.cwd(), 'README.md');
    
    if (!fs.existsSync(readmePath)) {
      console.error(`README.md not found at ${readmePath}`);
      return;
    }
    
    // Đọc nội dung README hiện tại
    let content = fs.readFileSync(readmePath, 'utf-8');
    
    // Tìm phần Discord hiện tại và thay thế
    const discordPattern = /### 🎮 Discord[\s\S]*?<!-- Thông tin này được cập nhật tự động từ API:.*?-->/;
    
    if (discordPattern.test(content)) {
      // Thay thế phần hiện có
      content = content.replace(discordPattern, discordSection);
    } else {
      // Thêm phần mới phía sau phần "Các Kỹ Năng Khác"
      if (content.includes("---\n\n### 📊 Thống Kê GitHub")) {
        content = content.replace("---\n\n### 📊 Thống Kê GitHub", `---\n\n${discordSection}\n\n---\n\n### 📊 Thống Kê GitHub`);
      } else {
        content = `${content}\n\n${discordSection}`;
      }
    }
    
    // Ghi lại nội dung đã cập nhật
    fs.writeFileSync(readmePath, content, 'utf-8');
    
    console.log("README.md updated with Discord presence");
    
  } catch (error) {
    console.error("Error updating README:", error);
  }
}

// Chạy script
updateReadme(); 